### ISSUE TITLE
How to load tfrecord on pytorch for multi-GPU training

### ISSUE THREAD

**Issue by mchl0203** [NVIDIA/DALI#1799]
I want to load tfrecord on pytorch for multi-GPU training, code using only one GPU is as follows:
```
class TFRecordPipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, num_gpus, tfrecord_file, tfrecord_idx):
        super(TFRecordPipeline, self).__init__(batch_size, num_threads, device_id)
        self.input = ops.TFRecordReader(path=tfrecord_file,
                                        index_path=tfrecord_idx,
                                        features={"image": tfrec.FixedLenFeature((), tfrec.string, ""),
                                                  "label": tfrec.FixedLenFeature([1], tfrec.int64, -1),
                                                  })
        self.decode = ops.ImageDecoder(device="mixed", output_type=types.BGR)  
        self.cmn = ops.CropMirrorNormalize(
            device="gpu",
            output_dtype=types.FLOAT,
            mean=[0.],
            std=[255.],
            output_layout="CHW")

    def define_graph(self):
        inputs = self.input(name="Reader")
        images = inputs["image"]
        label = inputs["label"].gpu()
        images = self.decode(images)
        images = self.cmn(images)
        return images, label

    def iter_setup(self):
        pass

pipes = [TFRecordPipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id, num_gpus=num_gpus,
                              tfrecord_file=tfrecord_file, tfrecord_idx=tfrecord_idx) for device_id in range(num_gpus)]
for i in range(num_gpus):
    pipes[i].build()

train_loader = DALIGenericIterator(pipes, ['image', 'label"], pipes[0].epoch_size("Reader"))
...

...
for epoch in range(args.end_epoch + 1):
    for batch_data in tqdm(train_loader,
                           total=math.ceil(train_loader._size / train_loader.batch_size) if use_dali else None):
        if use_dali:
            img = batch_data[0]["image"]
            label_gt = batch_data[0]["label"]
        label_pred = mymodel(img)
...
    train_loader.reset()
```

If I use 2 gpu, `batch_data[0]` and `batch_data[1]` have the same output, and the number of iterations of `train_loader` is half of `math.ceil(train_loader._size / train_loader.batch_size)`. 
So I modified the code.
```
train_loader = DALIGenericIterator(pipes, ['image', 'label"], pipes[0].epoch_size("Reader") * num_gpu)
...
if use_dali:
    img1 = batch_data[0]["image"]
    img2 = batch_data[1]["image"]
...
label_pred_1 = mymodel(img1)
label_pred_2 = mymodel(img2)
loss_1 = criterion(label_pred_1, label_gt)
loss_2 = criterion(label_pred_2, label_gt)
loss = (loss_1 + loss_2) / 2.0
loss.backward()
```
Is it correct that my code like this?
How should I modify my code, thank you very much ~

**Comment by JanuszL** [comment]
Hi,
If I understand your question correctly you need to set the `num_shards` and `shard_id` inside the `TFRecordReader`. If you are using 2 GPUs `num_shards = 2`, and one DALI instance should get `0` and the second one `1`. This way DALI will split data set into

**Comment by mchl0203** [comment]
@JanuszL Thank you for your reply! I use 2 GPUs, and set `num_shards = 2`, `shard_id = device_id`, `DALIGenericIterator(pipes, ['image', 'label"], pipes[0].epoch_size("Reader") * num_gpu)`. There is no problem with the number of pictures output if I use this way.

So at this time, there are 2 copies of data on 2 GPUs (data content may be different). I use `torch.nn.DataParallel` put the model on 2 GPUs, then each model is trained using the data of its own GPU, and the 2 GPUs do not interfere with each other. So my code below does not need to be modified. DALI uses 2 GPUs, which is equivalent to training my model twice. Is my understanding correct? Looking forward to your reply～
```
loss_1 = criterion(label_pred_1, label_gt)
loss_2 = criterion(label_pred_2, label_gt)
loss = (loss_1 + loss_2) / 2.0
loss.backward()
```

**Comment by JanuszL** [comment]
I think it is more or less aligned with this guide https://pytorch.org/tutorials/beginner/blitz/data_parallel_tutorial.html.
You can even do:
```
for d in batch_data:
    img1 = d["image"]
    output = model(img1)
```
Regarding the loo calculation - I'm not sure, you should better refer to PyTorch examples/guides like [this one](https://pytorch.org/tutorials/intermediate/model_parallel_tutorial.html).
Or ask on https://discuss.pytorch.org/.
Still, the recommended way to train on the GPU is distributed data parallel as it provides better load balancing between GPUs. You can refer to [NVIDIA official examples](https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/Classification/ConvNets/resnet50v1.5).

**Comment by mchl0203** [comment]
I don’t know much about parallel training, I am learning it, thank you very much for your reply!
